### PR TITLE
ONT-1017：add PreExecCheck in interface

### PR DESCRIPTION
**Pull request by xizho10**
1. add PreExecCheck in send transaction
2. add height in transaction json data 
3. add comment in interface for godoc

### CODE REVIEW

File: http/base/common/common.go [hunk 1]
```diff
@@ -521,3 +527,88 @@ func BuildNeoVMParam(builder *neovm.ParamsBuilder, smartContractParams []interfa
...
+	}
+	result, err := ledger.DefLedger.PreExecuteContract(txn)
+	if err != nil {
+		return fmt.Errorf("PreExecuteContract fail")
+	}
+	if txn.GasLimit < result.Gas {
+		return fmt.Errorf("GasLimit should >= %d", result.Gas)
+	}
+	gas, overflow := common.SafeMul(txn.GasPrice, result.Gas)
+	if overflow {
```
Comment (laizy) [inline http/base/common/common.go:556]: have checked in line 523.

File: http/base/common/common.go [hunk 2]
```diff
@@ -521,3 +527,88 @@ func BuildNeoVMParam(builder *neovm.ParamsBuilder, smartContractParams []interfa
...
+	if _, overflow := common.SafeMul(txn.GasLimit, txn.GasPrice); overflow {
+		return fmt.Errorf("Tx GasPrice %d and gasLimit %d safeMul overflow",
+			txn.GasPrice, txn.GasLimit)
+	}
+	if txn.TxType == types.Deploy && txn.GasLimit < svrneovm.CONTRACT_CREATE_GAS {
+		return fmt.Errorf("Deploy tx gasLimit should >= %d",
+			svrneovm.CONTRACT_CREATE_GAS)
+	}
+	gasLimitConfig := config.DefConfig.Common.GasLimit
+	gasPriceConfig := getGasPriceConfig()
```
Comment (laizy) [inline http/base/common/common.go:543]: i think you need cache this value and  refresh it  when it is expired, instead of fetching it every time.

Comment (laizy) [reply http/base/common/common.go:543]: given consensus node will disable pre-execute, performance is not crucial for client node, so keep it like this.

File: http/base/common/common.go (hunk 1 above)
Comment (xizho10) [reply http/base/common/common.go:556]: The two check gasLimit is different

Comment (laizy) [reply http/base/common/common.go:556]: result.Gas <= txn.GasLimit,  so if line 523 passed, it will guarantee not overflow

Comment (xizho10) [reply http/base/common/common.go:556]: line 523  for get gas, not only for check overflow.

**Review by laizy — APPROVED** [review]

**Pull request merged**